File: helpdeskspringboot/src/main/java/com/ez/service/UserService.java
package com.ez.service;

import com.ez.dto.ChangePassword;
import com.ez.dto.EditProfile;
import com.ez.exception.*;
import com.ez.entity.User;

import javax.mail.MessagingException;
import java.util.List;

public interface UserService {

    // find user by user id
    User findById(Long id) throws UserNotFoundException;

    // search users by page and based on the search criteria.
    // parameters:
    //  - pageNumber: page number
    //  - pageSize: page size
    //  - searchTerm: ID, email, firstName, lastName, phone
    //  - role: '', 'ROLE_CUSTOMER', 'ROLE_SUPPORTER', 'ROLE_ADMIN'
    //  - status: '', 'Active', 'Inactive'
    List<User> searchUsers(int pageNumber, int pageSize, String searchTerm, String role, String status);

    // find user by email
    User findUserByEmail(String email);

    User isInactiveUser(String email) throws InactiveUserException;

    // calculate total of users based on the search criteria.
    // based on this total of users we can calculate total pages
    long getTotalOfUsers(String searchTerm, String role, String status);

    // create new user
    User createUser(User user) throws MessagingException, EmailExistException;

    // update existing user
    User updateUser(User user) throws MessagingException, EmailExistException, UserNotFoundException;

    // update user profile
    User updateProfile(EditProfile editProfile) throws MessagingException, UserNotFoundException;

    // reset password in case user forgot his/her password
    void resetPassword(String email) throws MessagingException, EmailNotFoundException;
    void changePassword(ChangePassword changePassword) throws MessagingException, EmailNotFoundException, OldPasswordIsNotMatchException, NewPasswordIsNotMatchException;

}

File: helpdeskangular/src/app/service/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../environments/environment';
import { Observable } from 'rxjs';
import { User } from '../entity/User';
import { CustomHttpRespone } from '../entity/CustomHttpResponse';
import { ChangePassword } from '../entity/ChangePassword';
import { ResetPassword } from '../entity/ResetPassword';
import { EditProfile } from '../entity/EditProfile';

@Injectable({ providedIn: 'root' })
export class UserService {

  // 'http://localhost:8080'
  host = environment.apiUrl;

  // number of users per a page(default = 5)
  pageSize = environment.pageSize;

  constructor(private http: HttpClient) { }

  // get users by page and based on the search criteria
  searchUsers(page: number, searchTerm: string, role: string, status: string): Observable<User[]> {

    return this.http.get<User[]>(
      `${this.host}/user-search?page=${page}&size=${this.pageSize}&searchTerm=${searchTerm}&role=${role}&status=${status}`
    )
  }

  // calculate total of users for count total of pages
  getTotalOfUsers(searchTerm: string, role: string, status: string): Observable<number> {

    // ex: http://localhost:8080/total-of-users?searchTerm=""&role=""&status=""
    return this.http.get<number>(
      `${this.host}/total-of-users?searchTerm=${searchTerm}&role=${role}&status=${status}`
    );

  }

  // create new user
  public createUser(user: User): Observable<User> {
    return this.http.post<User>(`${this.host}/user-create`, user);
  }

  // edit existing user
  public editUser(user: User): Observable<User> {
    return this.http.put<User>(`${this.host}/user-edit`, user);
  }

  // reset password
  public resetPassword(resetPassword: ResetPassword): Observable<CustomHttpRespone> {

    return this.http.put<CustomHttpRespone>(`${this.host}/reset-password`, resetPassword);

  }

  // update profile
  public updateProfile(editProfile: EditProfile): Observable<User> {
    return this.http.put<User>(`${this.host}/edit-profile`, editProfile);
  }

  // change password
  public changePassword(changePassword: ChangePassword): Observable<CustomHttpRespone> {
    return this.http.put<CustomHttpRespone>(`${this.host}/change-password`, changePassword);
  }

  // find user by user id
  findById(id: number): Observable<User> {
    return this.http.get<User>(`${this.host}/user-list/${id}`);
  }

  // // find user by email
  // findUserByEmail(email: string): Observable<User> {
  //   return this.http.get<User>(`${this.host}/user-list/${email}`);
  // }

}
